### PR TITLE
fix: update Vault CRD during a boot upgrade

### LINKED ISSUE #5867: 'step boot vault' should update the vault CRD when there are changes
### Summary 

The vault CRD is programatically created by `step boot vault` but when a vault exists the command just skips and exists. It that case, it should check if there are any changes to the vault CRD (e.g. the docker image version update or some default settings enabled) and update the CRD.  

There is currently an issue during the boot update process when the vault version gets updated. This new version doesn't get propagated into the vault CRD.

### Expected behavior

The vault CRD should be updated when there are any changes.

### PR DESCRIPTION
**Pull request by hferentschik**
fixes #5867

- Uses now an insert or update approach for the Vault CRD
- Uses helm install directly, not the addon wrapper
- Tries to improve logging

### CODE REVIEW

File: pkg/kube/vault/vault.go [hunk 1]
```diff
@@ -481,3 +478,44 @@ func GetAuthSaName(vault v1alpha1.Vault) string {
 
 	return name.(string)
 }
+
+// CreateOrUpdateVault creates the specified Vault CRD if it does not exist or updates it otherwise.
+func CreateOrUpdateVault(vault *v1alpha1.Vault, vaultOperatorClient versioned.Interface, ns string) error {
```
Comment (ccojocar) [inline pkg/kube/vault/vault.go]: It would be nice to have a unit test for this function.

File: pkg/cmd/create/create_vault.go [hunk 2]
```diff
@@ -223,12 +217,19 @@ func (o *CreateVaultOptions) CreateVault(vaultOperatorClient versioned.Interface
...
+	if err != nil {
+		return errors.Wrap(err, "loading docker images from versions repository")
 	}
-	if kubeProvider == cloud.AWS || kubeProvider == cloud.EKS {
-		err = o.createVaultAWS(vaultOperatorClient, vaultName, kubeClient, clusterName, vaultAuthServiceAccount)
+	vaultCRD, err := kubevault.NewVaultCRD(kubeClient, vaultName, o.Namespace, images, vaultAuthServiceAccount, o.Namespace, o.SecretsPathPrefix)
+
+	err = o.setCloudProviderSpecificSettings(vaultCRD, kubeProvider, clusterName, o.Namespace, vaultAuthServiceAccount, kubeClient, vaultOperatorClient)
+	if err != nil {
+		return err
```
Comment (ccojocar) [inline pkg/cmd/create/create_vault.go]: It would be nice to wrap this error.

File: pkg/cmd/create/create_vault.go [hunk 3]
```diff
@@ -252,6 +253,49 @@ func (o *CreateVaultOptions) CreateVault(vaultOperatorClient versioned.Interface
...
+	if o.IsJXBoot() {
+		// only when executing locally via `jx boot` we expect to have the appropriate credentials to modify cloud resources
+		// the pipeline user won't have permissions to execute cloud provider apis
+		switch kubeProvider {
+		case cloud.GKE:
+			cloudProviderConfig, err = o.vaultGKEConfig(vaultCRD, vaultOperatorClient, o.BucketName, o.KeyringName, o.KeyName, kubeClient, clusterName, vaultAuthServiceAccount)
+		case cloud.AWS, cloud.EKS:
+			cloudProviderConfig, err = o.vaultAWSConfig(vaultCRD, vaultOperatorClient, kubeClient, clusterName, vaultAuthServiceAccount)
+		default:
+			err = errors.Errorf("Vault is not supported on cloud provider %s", kubeProvider)
```
Comment (ccojocar) [inline pkg/cmd/create/create_vault.go]: error message is no capitalised in Go

File: pkg/cmd/create/create_vault.go [hunk 4]
```diff
@@ -252,6 +253,49 @@ func (o *CreateVaultOptions) CreateVault(vaultOperatorClient versioned.Interface
...
+
+		cloudProviderConfig = kubevault.CloudProviderConfig{
+			Storage:           vaultCRD.Spec.Config["storage"].(kubevault.Storage),
+			Seal:              vaultCRD.Spec.Config["seal"].(kubevault.Seal),
+			UnsealConfig:      existingVaultCRD.Spec.UnsealConfig,
+			CredentialsConfig: existingVaultCRD.Spec.CredentialsConfig,
+		}
+	}
+	if err != nil {
+		return err
```
Comment (ccojocar) [inline pkg/cmd/create/create_vault.go]: nit: wrap this error

File: pkg/cmd/step/boot/step_boot_vault.go [hunk 5]
```diff
@@ -249,20 +235,43 @@ func (o *StepBootVaultOptions) Run() error {
 		return errors.Wrapf(err, "saving secrets location in ConfigMap %s in namespace %s", kube.ConfigMapNameJXInstallConfig, ns)
 	}
 
-	log.Logger().Infof("finding vault '%s' in namespace %s", requirements.Vault.Name, ns)
+	err = createVaultOptions.CreateOrUpdateVault(vaultOperatorClient, requirements.Vault.Name, provider)
+	if err != nil {
+		return err
```
Comment (ccojocar) [inline pkg/cmd/step/boot/step_boot_vault.go]: I would wrap this error.

File: pkg/cmd/step/boot/step_boot_vault.go [hunk 6]
```diff
@@ -249,20 +235,43 @@ func (o *StepBootVaultOptions) Run() error {
...
+	}
+	return nil
+}
+
+func (o *StepBootVaultOptions) installOperator(requirements *config.RequirementsConfig, ns string) error {
+	tag, err := o.vaultOperatorImageTag(&requirements.VersionStream)
+	if err != nil {
+		return err
```
Comment (ccojocar) [inline pkg/cmd/step/boot/step_boot_vault.go]: nit: wrap error

File: pkg/cmd/create/create_vault.go (hunk 3 above)
Comment (hferentschik) [reply pkg/cmd/create/create_vault.go]: I know, but is this not Vault as Vault the product? But I can change it

File: pkg/cmd/create/create_vault.go (hunk 2 above)
Comment (hferentschik) [reply pkg/cmd/create/create_vault.go]: I'll wrap all the errors ...

**Comment by hferentschik** [comment]
@ccojocar didn't realize that the linter will also now complain about spelling mistakes.

**Comment by hferentschik** [comment]
Closing this PR for now. I'll create a new one once the remaining issues are resolved.

**Pull request closed without merging**